File: php/oauth_request.php
<?php
/**
 * UserSpice OAuth Client - Authorization Request
 *
 * This script initiates the OAuth 2.0 Authorization Code flow.
 * It redirects the user to the UserSpice OAuth server for authentication.
 *
 * Flow:
 * 1. User visits this page
 * 2. Script generates a CSRF state token and stores it in session
 * 3. User is redirected to the OAuth server's authorization endpoint
 * 4. User logs in and authorizes the application
 * 5. OAuth server redirects back to oauth_response.php with auth code
 */

require_once 'oauth_config.php';

// Start session for state management
session_start();

// OAuth server authorization endpoint
$authEndpoint = $oSettings['server_url'] . 'users/auth/';

// Generate a random state parameter for CSRF protection
try {
    $state = bin2hex(random_bytes(16));
} catch (Exception $e) {
    die('Error: Unable to generate secure state parameter.');
}

// Store the state in session for verification in oauth_response.php
$_SESSION['oauth_state'] = $state;

// Build the authorization URL with required parameters
$authParams = [
    'response_type' => 'code',                    // We want an authorization code
    'client_id'     => $oSettings['client_id'],   // Identifies this application
    'redirect_uri'  => $oSettings['redirect_uri'], // Where to send the user after auth
    'state'         => $state,                     // CSRF protection token
    'scope'         => $oSettings['scope'],        // What permissions we're requesting
];

$authUrl = $authEndpoint . '?' . http_build_query($authParams);

// Redirect the user to the OAuth server for authentication
header('Location: ' . $authUrl);
exit;

File: java/OAuthRequest.java
// Java Vanilla OAuth Client Example for UserSpice
// Simulates initiating the OAuth request

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.UUID;

public class OAuthRequest {

    public static void main(String[] args) {
        Map<String, String> settings = OAuthClientConfig.getSettings();

        String authorizationEndpoint = settings.get("server_url") + settings.get("auth_endpoint_path");

        // Generate a random state parameter for CSRF protection
        String state = UUID.randomUUID().toString();
        // In a real web app, you would store this 'state' in the user's session
        // to verify it on callback.
        System.out.println("Generated state (should be stored in session): " + state);

        try {
            // Build the authorization URL
            String authUrl = authorizationEndpoint + "?" +
                    "response_type=code" +
                    "&client_id=" + URLEncoder.encode(settings.get("client_id"), StandardCharsets.UTF_8.name()) +
                    "&redirect_uri=" + URLEncoder.encode(settings.get("redirect_uri"), StandardCharsets.UTF_8.name()) +
                    "&state=" + URLEncoder.encode(state, StandardCharsets.UTF_8.name()) +
                    "&scope=" + URLEncoder.encode("profile", StandardCharsets.UTF_8.name()); // Add any scopes you need

            System.out.println("\n--- OAuth Authorization Request Simulation ---");
            System.out.println("In a real web application, you would redirect the user to the following URL:");
            System.out.println(authUrl);
            System.out.println("\nAfter the user authorizes, they will be redirected to your redirect_uri with a 'code' and 'state' parameter.");
            System.out.println("Example: " + settings.get("redirect_uri") + "?code=AUTHORIZATION_CODE_HERE&state=" + state);

        } catch (Exception e) {
            System.err.println("Error building authorization URL: " + e.getMessage());
            e.printStackTrace();
        }
    }
}
